gensim/models/doc2vec.py

```diff
@@ -585,18 +592,22 @@ def _vocab_from(self, documents, progress_per=10000):
                 interval_start = default_timer()
                 interval_count = total_words
             document_length = len(document.words)
+
             for tag in document.tags:
                 self.docvecs.note_doctag(tag, document_no, document_length)
+
             for word in document.words:
-                total_words += 1
-                if word in vocab:
-                    vocab[word].count += 1
-                else:
-                    vocab[word] = Vocab(count=1)
+                vocab[word] += 1
+
+            if self.max_vocab_size and len(vocab) > self.max_vocab_size:
+                total_words += utils.prune_vocab(vocab, min_reduce)
+                min_reduce += 1
```

Comment:
Or, stop escalating (as much) when reaching `min_count`. Basically linking `min_reduce` to `min_count` -- whatever ends up below `min_count` will get thrown away later anyway.

I should add this is all rather hypothetical: default is no pruning, and when a power users wants very large vocabulary (small `min_count`), they'll have enough RAM for very large `max_vocab_size` as well.

Now that I think of it, `max_vocab_size` may not be necessary at all -- the bottleneck is always `self.vocab`, with its size determined by `min_count`. Size of `self.raw_vocab` only comes into play with 1) super large corpora (tens of millions of distinct word types) AND 2) small RAM implying small `max_vocab_size` AND 3) large `min_count` for a small `self.vocab`. Only then does the pruning from 2) have a change to affect 3)...